crates/synthesis/src/discrete_basis/math.rs

```diff
@@ -0,0 +1,248 @@
+    let index = axis
+        .iter()
+        .enumerate()
+        .find(|(_, &el)| el.abs() > tol)
+        .expect("At least one element must be nonzero.")
+        .0;
+    match index {
+        0 => {
+            if matrix[(0, 1)] < 0. {
+                axis[1] *= -1.;
+            };
+            if matrix[(0, 2)] < 0. {
+                axis[2] *= -1.;
+            }
+        }
+        1 => {
+            if matrix[(1, 2)] < 0. {
+                axis[2] *= -1.;
+            }
+        }
+        _ => (),
+    };
+    axis
+}
+
+/// Compute the SO(3) matrix that rotates from ``from`` to ``to``.
+fn rotation_matrix(from: &Matrix3x1<f64>, to: &Matrix3x1<f64>, do_checks: bool) -> Matrix3<f64> {
+    let from = from.normalize();
+    let to = to.normalize();
+    let dot = from.dot(&to);
+
+    let mut cross = Matrix3::zeros();
+    add_cross_prod(1., &from.cross(&to), &mut cross);
+
+    let out = Matrix3::identity() + cross + cross * cross / (1. + dot);
+    if do_checks {
+        assert_so3("rotation matrix", &out);
+    }
+    out
+}
+
+/// Compute the SO(3) matrix implementing rotation about ``axis`` for an ``angle``.
+fn so3_from_angle_axis(angle: f64, axis: &Matrix3x1<f64>) -> Matrix3<f64> {
+    let mut out = Matrix3::<f64>::identity() * angle.cos();
+    add_cross_prod(angle.sin(), axis, &mut out);
+    add_outer_prod(1. - angle.cos(), axis, &mut out);
+
+    out
+}
+
+/// Decompose the SO(3) input matrix M into a balanced group commutator, that is
+///
+///     M = V W V^T W^T
+///
+/// for two SO(3) matrices V and W. See section 4.1 of https://arxiv.org/abs/quant-ph/0505030.
+pub fn group_commutator_decomposition(
+    matrix_so3: &Matrix3<f64>,
+    do_checks: bool,
+) -> (Matrix3<f64>, Matrix3<f64>) {
+    if do_checks {
+        assert_so3("GC input", matrix_so3);
+    }
+
+    let angle = solve_decomposition_angle(matrix_so3);
+
+    let e1 = Matrix3x1::new(1., 0., 0.);
+    let e2 = Matrix3x1::new(0., 1., 0.);
+
+    let vx = so3_from_angle_axis(angle, &e1);
+    let wy = so3_from_angle_axis(angle, &e2);
+
+    let tol = 1e-15;
+    let group_comm = group_commutator(&vx, &wy);
+    let group_comm_axis = rotation_axis_from_so3(&group_comm, tol);
+    let matrix_axis = rotation_axis_from_so3(matrix_so3, tol);
+
+    let sim_matrix = rotation_matrix(&group_comm_axis, &matrix_axis, do_checks);
+    let sim_matrix_t = sim_matrix.transpose();
+
+    let v = sim_matrix * vx * sim_matrix_t;
+    let w = sim_matrix * wy * sim_matrix_t;
+
+    if do_checks {
+        assert_so3("group commutator", &group_comm);
+        assert_so3("vx", &vx);
+        assert_so3("wy", &wy);
+        assert_so3("sim_matrix", &sim_matrix);
+        assert_so3("v", &v);
+        assert_so3("w", &w);
+    }
+
+    (v, w)
+}
+
+pub(super) fn assert_so3(name: &str, matrix: &Matrix3<f64>) {
+    if matrix.iter().any(|el| el.is_nan()) {
+        panic!("{} has NaN value.", name);
+    }
+    if (1. - matrix.determinant()) > (1e-5) {
```

Comment:
1e-5 is tol as defined below? how did you decide on this value? perhaps would be good do define it as a constant somewhere. also note that "tol" might be confusing with the "tol" parameter of SK.